src/validation.cpp

```diff
@@ -5369,3 +5553,62 @@ std::optional<int> ChainstateManager::getSnapshotBaseHeight()
     CBlockIndex* base = getSnapshotBaseBlock();
     return base ? std::make_optional(base->nHeight) : std::nullopt;
 }
+
+void ChainstateManager::reset()
+{
+    m_ibd_chainstate.reset();
+    m_snapshot_chainstate.reset();
+    m_active_chainstate = nullptr;
+    m_snapshot_validated = false;
+}
+
+void ChainstateManager::validatedSnapshotCleanup()
+{
+    AssertLockHeld(::cs_main);
+    auto get_storage_path = [](auto& chainstate) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) -> std::optional<fs::path> {
+        if (!(chainstate && chainstate->HasCoinsViews())) {
+            return {};
+        }
+        return chainstate->CoinsDB().StoragePath();
+    };
+    std::optional<fs::path> ibd_chainstate_path_maybe = get_storage_path(m_ibd_chainstate);
+    std::optional<fs::path> snapshot_chainstate_path_maybe = get_storage_path(m_snapshot_chainstate);
+    bool should_cleanup_snapshot = m_snapshot_chainstate && m_snapshot_validated;
+
+    if (!should_cleanup_snapshot) {
+        // No need to clean up.
+        return;
+    }
+    // If either path doesn't exist, that means at least one of the chainstates
+    // is in-memory, in which case we can't do on-disk cleanup. You'd better be
+    // in a unittest!
+    if (!ibd_chainstate_path_maybe || !snapshot_chainstate_path_maybe) {
+        LogPrintf("[snapshot] snapshot chainstate cleanup cannot happen with " /* Continued */
+                "in-memory chainstates - you are testing, right?\n");
+        return;
+    }
+
+    auto snapshot_chainstate_path = *snapshot_chainstate_path_maybe;
+    auto ibd_chainstate_path = *ibd_chainstate_path_maybe;
+
+    assert(fs::exists(snapshot_chainstate_path));
+    assert(fs::exists(ibd_chainstate_path));
+
+    LogPrintf("[snapshot] deleting background chainstate directory (now unnecessary) (%s)\n",
+        fs::PathToString(ibd_chainstate_path));
+
+    fs::path tmp_old{ibd_chainstate_path + "_todelete"};
+
+    fs::rename(ibd_chainstate_path, tmp_old);
+    fs::rename(snapshot_chainstate_path, ibd_chainstate_path);
+
+    LogPrintf("[snapshot] moving snapshot chainstate (%s) to " /* Continued */
+        "default chainstate directory (%s)\n",
+        fs::PathToString(snapshot_chainstate_path), fs::PathToString(ibd_chainstate_path));
```

Comment:
In commit "validation: add ChainMan logic for completing UTXO snapshot validation" (2efe7c7ec82e7885e07140e1deb5b5e9d6869f1d)

I'm suprised you can just rename directories while leveldb is still using them. If this works, I guess it's ok, but I think more ideally CChainstate objects would have `Destroy()` and `Rename(new_name)` methods and would do these operations in a safe way without ChainStateManager being involved at such a low level.